src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.Utf8.cs

```diff
@@ -0,0 +1,649 @@
+                // the case where the equality check is likely to succeed, which means that we want to avoid
+                // branching within this loop unless we're about to exit the loop, either due to failure or
+                // due to us running out of input data.
+
+                if (!Utf8Utility.UInt32OrdinalIgnoreCaseAscii(valueAu32, valueBu32))
+                {
+                    return false;
+                }
+
+                byteOffset += 4;
+                length -= 4;
+            }
+
+            if (length != 0)
+            {
+                // We have 1, 2, or 3 bytes remaining. We can't do anything fancy
+                // like backtracking since we could have only had 1-3 bytes. So,
+                // instead we'll do 1 or 2 reads to get all 3 bytes. Endianness
+                // doesn't matter here since we only compare if all bytes are ascii
+                // and the ordering will be consistent between the two comparisons
+
+                if (length == 3)
+                {
+                    valueAu32 = Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref charA, byteOffset));
+                    valueBu32 = Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref charB, byteOffset));
+
+                    byteOffset += 2;
+
+                    valueAu32 |= (uint)(Unsafe.AddByteOffset(ref charA, byteOffset) << 16);
+                    valueBu32 |= (uint)(Unsafe.AddByteOffset(ref charB, byteOffset) << 16);
+                }
+                else if (length == 2)
+                {
+                    valueAu32 = Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref charA, byteOffset));
+                    valueBu32 = Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref charB, byteOffset));
+                }
+                else
+                {
+                    Debug.Assert(length == 1);
+
+                    valueAu32 = Unsafe.AddByteOffset(ref charA, byteOffset);
+                    valueBu32 = Unsafe.AddByteOffset(ref charB, byteOffset);
+                }
+
+                if (!Utf8Utility.AllBytesInUInt32AreAscii(valueAu32 | valueBu32))
+                {
+                    // one of the inputs contains non-ASCII data
+                    goto NonAscii32;
+                }
+
+                if (lengthA != lengthB)
+                {
+                    // Failure if we reached the end of one, but not both sequences
+                    return false;
+                }
+
+                if (valueAu32 == valueBu32)
+                {
+                    // exact match
+                    return true;
+                }
+
+                if (!Utf8Utility.UInt32OrdinalIgnoreCaseAscii(valueAu32, valueBu32))
+                {
+                    return false;
+                }
+
+                byteOffset += 4;
+                length -= 4;
+            }
+
+            Debug.Assert(length == 0);
```

Comment:
I am quite confused as to how testing was good before and is failing now, but this snippet looks suspicious.  The length was 1-3, and byteOffset (though not used after this point) also would be increased by 1-3 (except for the 2 sometimes already added above)